runtime/autoload/provider/node.vim

```diff
@@ -66,9 +64,35 @@ function! provider#node#Detect() abort
   if !s:is_minimum_version(v:null, 6, 0)
     return ''
   endif
-  let entry_point = s:find_node_client('npm')
-  let entry_point = !empty(entry_point) ? entry_point : s:find_node_client('yarn')
-  return entry_point
+
+  let yarn_subpath = '/node_modules/neovim/bin/cli.js'
+  let npm_subpath = '/neovim/bin/cli.js'
+
+  " `yarn global dir` is slow (> 250ms), try the default path first
+  if filereadable('$HOME/.config/yarn/global' . yarn_subpath)
+      return '$HOME/.config/yarn/global' . yarn_subpath
+  end
+
+  " try both npm and yarn simultaneously
+  let yarn_opts = s:NodeHandler.new()
+  let yarn_opts.entry_point = yarn_subpath
+  let yarn_opts.job_id = jobstart(['yarn', 'global', 'dir'], yarn_opts)
+  let npm_opts = s:NodeHandler.new()
+  let npm_opts.entry_point = npm_subpath
+  let npm_opts.job_id = jobstart(['npm', '--loglevel', 'silent', 'root', '-g'], npm_opts)
+
+  " npm returns the directory faster, so let's check that first
+  let result = jobwait([npm_opts.job_id])
+  if npm_opts.result != ''
+      return npm_opts.result
+  endif
+
+  let result = jobwait([yarn_opts.job_id])
```

Comment:
`let result` is never used, so would be better to just not assign it.

    jobwait([yarn_opts.job_id])

Reply:
I just ended up using it to assert a zero exit code.